### DISCUSSION TITLE
Roadwarrior IP reservation

### DISCUSSION THREAD

**Discussion by ptchuba** [strongswan/strongswan#1514] [Q&A] [answered]
I'm trying to build a roadwarrior config such that each of my users receive a previously designated IP address. In other words, a user should receive the same IP always. I should be able to have 1000 users or more. 

Based on my research I'll either have to stand up a radius server (which i don't want to - another thing to worry about) or use IP Pool which requires attr-sql and sqlite plugins. I'm running Ubuntu 20.04 and so far have not been able to find a package that installs these plugins. 

Since I couldn't find any, i followed the instructions and enabled the options with ./configure. I do not get any errors with make install but a client that was able to connect with IPv6 previously (when i used apt package install) now causes the server to crash when they try to connect. I don't want to troubleshoot this, just adding here for context.

So my question is, without recompiling is there any other way I can get these plugins installed and enabled and/or is there any other method to assign specific IP addresses to clients (without separate conns of course)?

Thanks

**Comment by tobiasbrunner** [comment, marked as answer]
> Based on my research I'll either have to stand up a radius server (which i don't want to - another thing to worry about) or use IP Pool which requires attr-sql and sqlite plugins.

Another alternative is the _dhcp_ plugin (Ubuntu ships it in the _libcharon-extra-plugins_ package), see [the docs](https://docs.strongswan.org/docs/5.9/features/vip.html#_responder_configuration).

> I don't want to troubleshoot this, just adding here for context.

You might be mixing plugins/libs from different versions/builds. Have you removed all binary packages before installing the custom built binaries?

> So my question is, without recompiling is there any other way I can get these plugins installed and enabled and/or is there any other method to assign specific IP addresses to clients (without separate conns of course)?

I guess you don't want to rebuild the Ubuntu packages (after changing the configure options in `debian/rules` etc.)? You could theoretically compile the plugins if you use the **exact** same source and build options/environment and manually install just them together with the binary packages.

**Reply by ptchuba** [reply]
Thank you so much for the reply. You were right about mixing plugins/libs. I started from a clean slate and compiled the plugins I needed as well as the default ones if I was using the Ubuntu binary installation. Everything now works as it used to. I added the configuration of the attr-sql plugin, along with sqlite setup as shown [here](https://www.strongswan.org/uml/testresults/ikev2/ip-pool-db/index.html). 

I then created a pool using the `--addresses <file>` option of ipsec pool. The file just lists the identities and the IPs as follows
```
172.25.0.60=user2@example.local
172.25.0.65=user5@example.local
172.25.0.70=user7@example.local
172.25.0.80=user8@example.local
```
However, when if I connect user8 first, they get the fist IP address of the pool, i.e. the pool works but the reservation doesn't. I've tried changing the identity to `userx\@example.local` as seen on one of the examples in the docs but it doesn't work. 

Logs don't have any errors, just shows assignment
```
Jan 19 06:00:45 strongswan-vdi-ep charon: 14[CFG] acquired new lease for address 172.25.0.60 in pool 'sqlite_pool'
Jan 19 06:00:45 strongswan-vdi-ep charon: 14[IKE] assigning virtual IP 172.25.0.60 to peer 'user8@example.local'
```

Please advise.

**Reply by tobiasbrunner** [reply]
Please post the output of `ipsec pool --status` and `ipsec pool --leases` (the latter after a client connected).

> I've tried changing the identity to `userx\@example.local` as seen on one of the examples in the docs but it doesn't work.

Sounds like a typo, where did you see that?

**Reply by ptchuba** [reply]
> Sounds like a typo, where did you see that?

https://docs.strongswan.org/docs/5.9/tools/pool.html in the section `Replace an existing pool in the database`

Thank you so much for your help with this and sorry for the delay in responding. I've since made some progress. With my linux clients running strongswan I'm now able to assign the IPs correctly using either the attr-sql plugin or dhcp plugin (and running a dhcp server - dnsmasq). However, the Windows boxes are a different story. I noticed and confirmed from different posts that Strongswan sends the Subject DN instead as the client ID. 

1. Using DHCP plugin with dnsmasq. I need to be able to pre-determine the identity that is sent by Windows so that I can reserve an ip in the dnsmasq config file. Without any reservations, Windows gets an IP and I see this in the Dnsmasq's lease file.

```
root@strongswan-vdi-ep:/home/admin# cat /var/lib/misc/dnsmasq.leases 
1674601869 7a:a7:5b:59:2b:60 172.25.0.44 * 30:5b:31:1c:30:1a:06:03:55:04:03:0c:13:75:73:65:72:31:40:65:78:61:6d:70:6c:65:2e:6c:6f:63:61:6c:31:0b:30:09:06:03:55:04:06:13:02:55:53:31:0f:30:0d:06:03:55:04:07:0c:06:41:75:73:74:69:6e:31:10:30:0e:06:03:55:04:0a:0c:07:45:78:61:6d:70:6c:65:31:0b:30:09:06:03:55:04:0b:0c:02:49:54
```
The subject DN of the windows certificate is `CN = user1@example.local, C = US, L = Austin, O = Example, OU = IT` and this is what I see in the authentication logs on Strongswan. I tried converting the entire subject DN to Hex and I get a portion of the hex string above but not exactly the same and using what I get to reserve the IP does not work. However if i used the Hex string above in dnsmasq to reserve the IP, it works just fine. So my issue is how to convert the Subject DN to the hex string above.  Or rather, how does Strongswan convert it to add to Option 61 in the DHCP Discover packet?

2. Alternatively Using attr-sql plugin and IPSec Pool. When I create a pool with available addresses, the Windows box gets an IP and I see the Subject DN in the identity section as `CN = user1@example.local, C = US, L = Austin, O = Example, OU = IT`. I'm using the  `--addresses <file>` option and I've tried adding the subject DN to the file with no luck. I've tried it with or without double quotes, single quotes, square brackets and just doesn't work i.e. 

```
172.25.0.80="CN = user1@example.local, C = US, L = Austin, O = Example, OU = IT"
```
or 
```
172.25.0.80='CN = user1@example.local, C = US, L = Austin, O = Example, OU = IT'
```
How do I format this correctly?

Also, do I need to reload strongswan for each change to the IPsec pool configuration? I need to be able to make changes frequently without affecting existing connections.

Thanks
Peter

**Reply by tobiasbrunner** [reply]
> https://docs.strongswan.org/docs/5.9/tools/pool.html in the section `Replace an existing pool in the database`

Thanks, fixed.

> I noticed and confirmed from different posts that Strongswan sends the Subject DN instead as the client ID.

Do you mean Windows? Because strongSwan can send either the full subject DN (which is the default if a certificate but no identity is configured) or use a subjectAltName as identity as well. But I guess Windows will always send the full subject DN.

> Or rather, how does Strongswan convert it to add to Option 61 in the DHCP Discover packet?

strongSwan just uses the exact encoding of the identity sent by the client, which looks just like the literal ASN.1 encoding of the subject DN of the certificate (`0x30` is how a ASN.1 SEQUENCE starts). If you have a certificate, you can use the [`pki --dn --format hex`](https://docs.strongswan.org/docs/5.9/pki/pkiDn.html) command to get a hex encoding of the subject DN.

While the [`id2sql` script](https://github.com/strongswan/strongswan/blob/master/scripts/id2sql.c) could also be used to convert an identity string to a binary representation, the problem is that the binary representation of such a DN might not be the exact same. For instance, the  DN above apparently uses `UTF8String` as type for all RDNs except `C`, for which `PrintableString` is used. However, strongSwan, when encoding the string `CN=user1@example.local, C=US, L=Austin, O=Example, OU=IT` as ASN.1 identity, will use `UTF8String` only for `CN` and for all other RDNs `PrintableString`. So the binary encoding won't match exactly. This doesn't matter when strongSwan itself matches identities as it basically ignores the type (e.g. if you configured the DN as remote identity in the config it would get matched successfully against this certificate). But that's not the case when binary encodings are compared by the DHCP server or via SQL. So you have to make sure that you use the actual binary encoding of the identity as seen in the certificate.

> How do I format this correctly?

The value after the first `=` is just parsed as is, so definitely don't use any quotes. However, the problem mentioned above will occur when the string is parsed and converted to a binary value that's stored in the database. So what you have to do is provide the actual binary encodings of these DNs. This should work:

```
172.25.0.80=asn1dn:#305b311c301a06035504030c137573657231406578616d706c652e6c6f63616c310b3009060355040613025553310f300d06035504070c0641757374696e3110300e060355040a0c074578616d706c65310b3009060355040b0c024954
```

**Reply by ptchuba** [reply]
Wow, this is great information, thank you so much!! My last challenge is figuring out how to get ipv6 addresses through dhcp. The dhcp plugin documentation does not mention v6 and I found an old post from 8 years ago  indicating it does not support dhcpv6. Is this still the case? 

Thanks

**Reply by tobiasbrunner** [reply]
> The dhcp plugin documentation does not mention v6 and I found an old post from 8 years ago indicating it does not support dhcpv6. Is this still the case?

Unfortunately, yes.

Note that Windows clients don't install routes for IPv6, [two client-side workarounds are documented](https://docs.strongswan.org/docs/5.9/interop/windowsClients.html#_ipv6). There is apparently an alternative that uses Router Advertisements sent from the server to the client (might work with [unicast RAs](https://wiki.strongswan.org/issues/817#note-13), which requires configuring the client's virtual IP in e.g. `radvd`, or with [multicast RAs](https://github.com/strongswan/strongswan/issues/315#issuecomment-1114229737) - the latter might be tricky with multiple clients connecting concurrently, though).

**Reply by ptchuba** [reply]
Thanks a lot @tobiasbrunner !! Very much appreciate the work you do!
